kubernetes/structures_test.go

```diff
@@ -35,3 +42,248 @@ func TestIsInternalKey(t *testing.T) {
 		})
 	}
 }
+
+func TestFlattenMetadataFields(t *testing.T) {
+	t.Parallel()
+
+	annotations := map[string]string{
+		"fake.kubernetes.io": "fake",
+	}
+	labels := map[string]string{
+		"foo": "bar",
+	}
+	uid := "7e9439cb-2584-4b50-81bc-441127e11b26"
+	cases := map[string]struct {
+		meta     metav1.ObjectMeta
+		expected []interface{}
+	}{
+		"DefaultNamespaceStaticName": {
+			metav1.ObjectMeta{
+				Annotations:     annotations,
+				GenerateName:    "",
+				Generation:      1,
+				Labels:          labels,
+				Name:            "foo",
+				Namespace:       "",
+				ResourceVersion: "1",
+				UID:             types.UID(uid),
+			},
+			[]interface{}{map[string]interface{}{
+				"annotations":      annotations,
+				"generation":       int64(1),
+				"labels":           labels,
+				"name":             "foo",
+				"resource_version": "1",
+				"uid":              uid,
+			}},
+		},
+		"NonDefaultNamespaceStaticName": {
+			metav1.ObjectMeta{
+				Annotations:     annotations,
+				GenerateName:    "",
+				Generation:      1,
+				Labels:          labels,
+				Name:            "foo",
+				Namespace:       "Test",
+				ResourceVersion: "1",
+				UID:             types.UID(uid),
+			},
+			[]interface{}{map[string]interface{}{
+				"annotations":      annotations,
+				"generation":       int64(1),
+				"labels":           labels,
+				"name":             "foo",
+				"namespace":        "Test",
+				"resource_version": "1",
+				"uid":              uid,
+			}},
+		},
+		"DefaultNamespaceGeneratedName": {
+			metav1.ObjectMeta{
+				Annotations:     annotations,
+				GenerateName:    "gen-foo",
+				Generation:      1,
+				Labels:          labels,
+				Name:            "",
+				Namespace:       "",
+				ResourceVersion: "1",
+				UID:             types.UID(uid),
+			},
+			[]interface{}{map[string]interface{}{
+				"annotations":      annotations,
+				"generate_name":    "gen-foo",
+				"generation":       int64(1),
+				"labels":           labels,
+				"name":             "",
+				"resource_version": "1",
+				"uid":              uid,
+			}},
+		},
+		"NonDefaultNamespaceGeneratedName": {
+			metav1.ObjectMeta{
+				Annotations:     annotations,
+				GenerateName:    "gen-foo",
+				Generation:      1,
+				Labels:          labels,
+				Name:            "",
+				Namespace:       "Test",
+				ResourceVersion: "1",
+				UID:             types.UID(uid),
+			},
+			[]interface{}{map[string]interface{}{
+				"annotations":      annotations,
+				"generate_name":    "gen-foo",
+				"generation":       int64(1),
+				"labels":           labels,
+				"name":             "",
+				"namespace":        "Test",
+				"resource_version": "1",
+				"uid":              uid,
+			}},
+		},
+	}
+	for _, c := range cases {
+		out := flattenMetadataFields(c.meta)
+		if !reflect.DeepEqual(out, c.expected) {
+			t.Fatalf("Error matching output and expected: %#v vs %#v", out, c.expected)
+		}
+	}
```

Comment:
nit: we could use `t.Run(<test name>, func(t *testing.T) { ... })` here to make the test cases explicit sub tests (ditto for other flattener unit tests)

Reply:
Thanks! Fixed.